### PR TITLE
Add an insecure option for testing hawkular using uncertified ssl connection

### PR DESCRIPTION
**Pull request by yaacov**
**Description**
When using hawkular python client using my python2 and urllib2 I was unable to connect to
the hawkular server because it was insecure. This PR adds the option to connect to an insecure server using ssl for development and testing.

**Example**
Without `context`:

```
client = HawkularMetricsClient(host='yzamir-centos7-1.eng.lab.tlv.redhat.com', port=443, scheme='https', token='...', tenant_id='_system')
>>> client.query_tenants()
 ... <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:590)>
```

With `context = ssl._create_unverified_context()`:

```
client = HawkularMetricsClient(host='yzamir-centos7-1.eng.lab.tlv.redhat.com', port=443, scheme='https', token='...', tenant_id='_system', context=ssl._create_unverified_context())
>>> client.query_tenants()
[{u'id': u'default'}, {u'id': u'openshift-infra'}, {u'id': u'_system'}]
```

### CODE REVIEW

**Comment by yaacov** [comment]
@simon3z @pilhuhn @lucasponce please review

**Comment by jpkrohling** [comment]
I'm against this change. Adding an "insecure" option will eventually lead to have a code in production with this value set to true, which completely defeats the purpose of using SSL in the first place. You should instead really try to get a proper certificate for your server.

**Comment by jpkrohling** [comment]
By the way: ping me on #hawkular and I can show you some alternatives :)

File: hawkular/metrics.py [hunk 1]
```diff
@@ -124,7 +127,13 @@ def _clean_metric_id(metric_id):
         return quote(metric_id, '')
 
     def _get_base_url(self):
-        return "{0}://{1}:{2}/{3}/".format(self.scheme, self.host, str(self.port), self.path)
+        scheme = self.scheme.lower()
```
Comment (simon3z) [inline hawkular/metrics.py]: @yaacov is this entire block related to your change or is just cosmetic?

File: hawkular/metrics.py [hunk 2]
```diff
@@ -177,7 +186,13 @@ def _http(self, url, method, data=None):
                     req.data = data.encode('utf-8')
 
             req.get_method = lambda: method
-            res = urlopen(req)
+
+            if self.insecure:
+                context = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
+                res = urlopen(req, context = context)
+            else:
+                res = urlopen(req)
```
Comment (simon3z) [inline hawkular/metrics.py]: @yaacov would this work?

``` ruby
context = ssl.SSLContext(ssl.PROTOCOL_TLSv1) if self.insecure
res = urlopen(req, context = context)
```

File: hawkular/metrics.py (hunk 1 above)
Comment (yaacov) [reply hawkular/metrics.py]: I do not know if it's a bug in urllib2 or the way it should be, on my system:

```
urlopen('https://example.com/foo') - works
urlopen('https://example.com:443/foo') - not working

```

If it's just on my system, I will remove this ( and have a locale version with it for me )
it it's on all systems with python2 and urllib2, I think it should stay.

Can anyone verify that it's just on my system or something wider ?

File: hawkular/metrics.py (hunk 2 above)
Comment (yaacov) [reply hawkular/metrics.py]: @simon3z Thanks, it works :-)

File: hawkular/metrics.py (hunk 1 above)
Comment (simon3z) [reply hawkular/metrics.py]: @yaacov that seems pretty weird.

Comment (yaacov) [reply hawkular/metrics.py]: @simon3z
I is :-) , I will try on my centos, and if it's just on my fedora i will remove it.

p.s. if @jpkrohling will show me another way to do this, i will close this PR.

**Comment by lucasponce** [comment]
Adding @burmanm to the review of this PR.
He is the lead for this client in particular.

**Comment by yaacov** [comment]
> Adding an "insecure" option will eventually lead to have a code in production with this value set to true

It makes my life while developing easier, But as a developer I can also have my own version of this client :-) 
It helps me, maybe it can help someone else, but then as @jpkrohling say it may cause more truble ... @burmanm WDYT ?

**Comment by jpkrohling** [comment]
The right thing to do isn't always easy ;)

**Comment by simon3z** [comment]
> I'm against this change. Adding an "insecure" option will eventually lead to have a code in production with this value set to true, which completely defeats the purpose of using SSL in the first place. You should instead really try to get a proper certificate for your server.

@jpkrohling every client using ssl supports this override (making clear that it's not for production).
Also, SSL has multiple purposes, for example encryption and data integrity would be still working.

**Comment by jpkrohling** [comment]
Neither encryption nor integrity can be guaranteed if the certificate is not trusted. If an encrypted payload can be intercepted, changes are extremely high that a MITM-attack can be performed.

That said, there are possible ways to provide a similar feature in a more secure way. One solution is to allow the client to provide a SSL Verifier, so that at least a certificate pinning is available. The way this patch is sent, it's an "either/or" approach, which is not good.

But really: the real solution is to get a valid cert. I provided a link to @yaacov to a service that Red Hat provides to internal services to get trusted certs. A secondary solution is to add the self-generated cert to the system's trusted certificates as part of the environment setup.

**Comment by yaacov** [comment]
@jpkrohling When writing code you can do things that are way stranger than trusting an insecure ssl connection. We allow developers to write code and we trust them not to do things too weird. I think we can trust an average developer not to write `insecure=True` in production.

But then again you may be right :-)

**Comment by burmanm** [comment]
The approach I've taken in the Golang client is that the client can push any sort of SSLContext they wish to the client and use that (and as such, move insecure handling to their own code - so they can use it if they want to, but we don't actually support it).

I think I'd like that approach as well, as the insecure is quite handy for some automated unit tests or integration tests. The urlopen(.., context) is however problematic as that syntax was introduced in Python 2.7.9, yet RHEL7 ships with Python 2.7.5. As such, I can't accept the current code since I need something that works out-of-the-box for RHEL7 also (and supports Python 3). So giving context in the NewHawkularClient would have otherwise been fine, if you can find a way to make it work for everyone (we could also remove the cafile / capath in that case from the init and only support setting context).

**Comment by yaacov** [comment]
> The urlopen(.., context) is however problematic as that syntax was introduced in Python 2.7.9, yet RHEL7 ships with Python 2.7.5.

@burmanm On my centos 7 with python 2.7.5, it seems to work, am I missing something ?

```
# python
Python 2.7.5 (default, Aug 18 2016, 15:58:25) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-4)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from urllib2 import Request, urlopen, URLError, HTTPError, HTTPErrorProcessor, build_opener, install_opener
>>> import ssl
>>> context = ssl._create_unverified_context()
>>> req = urlopen('https://www.google.com', context = context)
>>> req.read()
```

File: hawkular/metrics.py (hunk 1 above)
Comment (yaacov) [reply hawkular/metrics.py]: @simon3z checked on my centos, this block is unnecessary, removing it.
I do not know what is the problem on my system, but it is only on my system :-(

**Comment by burmanm** [comment]
I wonder if they've backported in that case as documentation states: "Changed in version 2.7.9: cafile, capath, cadefault, and context were added."

**Comment by yaacov** [comment]
> The approach I've taken in the Golang client is that the client can push any sort of SSLContext they wish

@burmanm @jpkrohling changed the PR to use `context` as the outside var :-)

**Comment by jpkrohling** [comment]
@yaacov , it looks really good now !
